ext/semian/simple_integer.c

```diff
@@ -0,0 +1,157 @@
+#include "simple_integer.h"
+
+#include "sysv_semaphores.h"
+#include "sysv_shared_memory.h"
+#include "types.h"
+#include "util.h"
+
+void
+semian_simple_integer_dfree(void* ptr)
+{
+  semian_simple_integer_t* res = (semian_simple_integer_t*)ptr;
+  free_shared_memory(res->shmem);
+}
+
+size_t
+semian_simple_integer_dsize(const void* ptr)
+{
+  return sizeof(semian_simple_integer_t);
+}
+
+static const rb_data_type_t semian_simple_integer_type = {
+  .wrap_struct_name = "semian_simple_integer",
+  .function = {
+    .dmark = NULL,
+    .dfree = semian_simple_integer_dfree,
+    .dsize = semian_simple_integer_dsize,
+  },
+  .data = NULL,
+  .flags = RUBY_TYPED_FREE_IMMEDIATELY,
+};
+
+static void init_fn(void* ptr)
+{
+  semian_simple_integer_shared_t* res = (semian_simple_integer_shared_t*)ptr;
+  res->val = 0;
+}
+
+void
+Init_SimpleInteger()
+{
+  dprintf("Init_SimpleInteger\n");
+
+  VALUE cSemian = rb_const_get(rb_cObject, rb_intern("Semian"));
+  VALUE cSimple = rb_const_get(cSemian, rb_intern("Simple"));
+  VALUE cSimpleInteger = rb_const_get(cSimple, rb_intern("Integer"));
+
+  rb_define_alloc_func(cSimpleInteger, semian_simple_integer_alloc);
+  rb_define_method(cSimpleInteger, "initialize_simple_integer", semian_simple_integer_initialize, 1);
+  rb_define_method(cSimpleInteger, "increment", semian_simple_integer_increment, -1);
+  rb_define_method(cSimpleInteger, "reset", semian_simple_integer_reset, 0);
+  rb_define_method(cSimpleInteger, "value", semian_simple_integer_value_get, 0);
+  rb_define_method(cSimpleInteger, "value=", semian_simple_integer_value_set, 1);
+}
+
+VALUE
+semian_simple_integer_alloc(VALUE klass)
+{
+  semian_simple_integer_t *res;
+  VALUE obj = TypedData_Make_Struct(klass, semian_simple_integer_t, &semian_simple_integer_type, res);
+  return obj;
+}
+
+VALUE
+semian_simple_integer_initialize(VALUE self, VALUE name)
+{
+  semian_simple_integer_t *res;
+  TypedData_Get_Struct(self, semian_simple_integer_t, &semian_simple_integer_type, res);
+  res->key = generate_key(to_s(name));
+
+  dprintf("Initializing simple integer '%s' (key: %lu)", to_s(name), res->key);
+  res->sem_id = initialize_single_semaphore(res->key, SEM_DEFAULT_PERMISSIONS);
+  res->shmem = get_or_create_shared_memory(res->key, &init_fn);
```

Comment:
There was a brief period where there was something else in the shared memory (I can't really remember now). There's no real issue with using a semaphore, and it would get rid of the need for locks, assuming we could guarantee that the operations would not block.

Reply:
I'll think more about this when I review the actual code and not just the approach.